### PR TITLE
451006-Snitavets Mikhail

### PR DESCRIPTION
**Pull request by Meosit**

### CODE REVIEW

File: Tracer/MethodNode.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+            MethodName = methodName;
+            ClassName = className;
+            ParameterCount = paramCount;
+            _startTime = DateTime.Now;
+            _endTime = DateTime.Now;
+        }
+
+        internal void StopTrace()
+        {
+            _endTime = DateTime.Now;
```
Comment (RamanBut-Husaim) [inline Tracer/MethodNode.cs]: @Meosit Try to use `DateTime.UtcNow` in internal calculations.

File: Tracer/MethodNode.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            InnerMethods = new List<MethodNode>();
+            MethodName = methodName;
+            ClassName = className;
+            ParameterCount = paramCount;
+            _startTime = DateTime.Now;
+            _endTime = DateTime.Now;
+        }
+
+        internal void StopTrace()
```
Comment (RamanBut-Husaim) [inline Tracer/MethodNode.cs]: @Meosit Do you think that it is logically correct to place a method `StopTrace` in this class?

File: Tracer/ThreadNode.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        private readonly Stack<MethodNode> _methodStack;
+        private MethodNode _currentNode;
+
+        private readonly DateTime _startTime;
+        private DateTime _endTime;
+        public TimeSpan ExecutionTime => _endTime - _startTime;
+
+        internal ThreadNode()
+        {
+            _startTime = DateTime.Now;
```
Comment (RamanBut-Husaim) [inline Tracer/ThreadNode.cs]: @Meosit `DateTime.UtcNow`

File: Tracer/ThreadNode.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            }
+            _currentNode = node;
+        }
+
+        internal void StopTraceMethod()
+        {
+            _currentNode.StopTrace();
+            if (_methodStack.Count == 0)
+            {
+                _endTime = DateTime.Now;
```
Comment (RamanBut-Husaim) [inline Tracer/ThreadNode.cs]: @Meosit `DateTime.UtcNow`

File: Tracer/XmlTraceResultFormatter.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+using System.Xml.Linq;
+
+namespace Tracer
+{
+    public class XmlTraceResultFormatter: ITraceResultFormatter
+    {
+
+        private string _path;
+
+        private bool IsValidPath(string path)
```
Comment (RamanBut-Husaim) [inline Tracer/XmlTraceResultFormatter.cs]: @Meosit No usages as far as I see.

File: Tracer/Tracer.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            lock (Lock)
+            { 
+                _traceResult.StopTraceMethod(Thread.CurrentThread.ManagedThreadId);
+            }
+           
+        }
+
+        public TraceResult GetTraceResult()
+        {
+            return _traceResult;
```
Comment (RamanBut-Husaim) [inline Tracer/Tracer.cs]: @Meosit There is a potential issue with this stuff. We'll discuss it on the lesson.

**Comment by Meosit** [comment]
Passed

**Comment by RamanBut-Husaim** [comment]
@Meosit Merged, thank you.

**Pull request merged**